src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+
+        /// <summary>Base type for all generator implementations that plug into the base Random.</summary>
+        private abstract class ImplBase
         {
-            if (minValue > maxValue)
-            {
-                throw new ArgumentOutOfRangeException(nameof(minValue), SR.Format(SR.Argument_MinMaxValue, nameof(minValue), nameof(maxValue)));
-            }
+            public abstract double Sample();
 
-            long range = (long)maxValue - minValue;
-            return range <= int.MaxValue ?
-                (int)(Sample() * range) + minValue :
-                (int)((long)(GetSampleForLargeRange() * range) + minValue);
-        }
+            public abstract int Next();
 
-        public virtual double NextDouble() => Sample();
+            public abstract int Next(int maxValue);
 
-        public virtual void NextBytes(byte[] buffer)
+            public abstract int Next(int minValue, int maxValue);
+
+            public abstract long NextInt64();
+
+            public abstract long NextInt64(long maxValue);
+
+            public abstract long NextInt64(long minValue, long maxValue);
+
+            public abstract double NextDouble();
+
+            public abstract float NextSingle();
+
+            public abstract void NextBytes(byte[] buffer);
+
+            public abstract void NextBytes(Span<byte> buffer);
+        }
+
+        #region Implementations
+        /// <summary>
+        /// Provides an implementation of the xoshiro256** algorithm. This implementation is used
+        /// when no seed is specified and an instance of the base Randomclass is constructed.  As
+        /// such, we are free to implement however we see fit, withoutback compat concerns around
+        /// the sequence of numbers generated or what methods call what other methods.
+        /// </summary>
+        private sealed class Xoshiro256StarStarImpl : ImplBase
         {
-            if (buffer is null)
+            // NextUInt64 is based on the algorithm from http://prng.di.unimi.it/xoshiro256starstar.c:
+            //
+            //     Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+            //
+            //     To the extent possible under law, the author has dedicated all copyright
+            //     and related and neighboring rights to this software to the public domain
+            //     worldwide. This software is distributed without any warranty.
+            //
+            //     See <http://creativecommons.org/publicdomain/zero/1.0/>.
+
+            private ulong _s0, _s1, _s2, _s3;
+
+            public unsafe Xoshiro256StarStarImpl()
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.buffer);
+                do
+                {
+                    Guid g;
+
+                    // Seed the four ulongs (4 64-bit values) from two Guids (2 128-bit values)
+                    g = Guid.NewGuid();
```

Comment:
Are there any concerns here with `Guid` itself not being random?

Reply:
This claims it is. https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Guid.Unix.cs#L16

Reply:
I went down the rabbit hole a bit and satisfied myself that the Windows ones do create cryptographically secure values (at least except for 6 bits indicating the version)

Reply:
It can call `Interop.GetRandomBytes` like the previous implementation that will avoid the problem with the 6 predictable bits in Guid.

Reply:
There are a couple of issues to bear in mind here, related to initialisation of the xoshiro/xoroshiro family of PRNGs:

1. The bits of the first sample from the PRNG are similar to the bits of the seed state. 
2. The seed state must not be all zeros.

Point 2 is specified in the comment at the top of the C reference implementation, here (http://prng.di.unimi.it/xoshiro256starstar.c):

>    The state must be seeded so that it is not everywhere zero. If you have
>    a 64-bit seed, we suggest to seed a splitmix64 generator and use its
>    output to fill s.

I don't recall where I read point 1, but I am certain of it (I can search for evidence, or we could just ask Sebastiano to be honest. Perhaps I read it in the research paper actually). The point is that there is insufficient mixing of bits going on before the first call/iteration of the PRNG, to break the relationship between the seed bits and the sample bits.

This means that if you e.g. use new Guid() for seed state, and some of the bits aren't random (which is the case) then the first sample will exhibit a strong bias towards certain values. 

On point 2, a crypto random seed is highly unlikely to be all zero of course, but it /could/ be theoretically. More practically, if we did ever allow the user's to init XoshiroStarStar256Impl with their own seed, then I have often seen `new Random(0)` in the wild! (and I confess that I have done that myself in unit tests!).

There is also this note on the xoshiro/xoroshiro home page (http://prng.di.unimi.it/):

> We suggest to use SplitMix64 to initialize the state of our generators starting from a 64-bit seed, as research has shown that initialization must be performed with a generator radically different in nature from the one initialized to avoid correlation on similar seeds.

Hmm, well, if you have a crypto random seed value then all is well (except for the theoretical 'all zeros' case), but for user supplied seeds we should hash the provided seed rather than directly storing it into the internal state variables.

Spitmix also has the property that it is [equidistributed](https://en.wikipedia.org/wiki/Equidistributed_sequence). In practical terms this just means it's as if we took the sequence (1, 2, 3... Int64.MaxValue) and shuffled the values, i.e. each value occurs once and once only - before looping back to repeat the sequence (as all PRNGs do eventually). This means that if we sample a zero, then the other three samples (we need for 64 bit sample for initial state) must be non-zero, thus guaranteeing that no matter what seed state was passed in, the state variables aren't all zero.

So Splitmix has been used to hash the passed in seed to break any correlation with the bits of that seed, and also to guarantee non-zero initial PRNG state.

Given that the PRNG author's recommend using Splitmix, I would advise sticking to that recommendation.

Sorry for raising all these little edge cases, but getting PRNGs right is quite hard it seems! :(

Reply:
Ahh, just noticed that the code does already guarantee non-zero state:

```C#
  do
  {
      Guid g;
  
      // Seed the four ulongs (4 64-bit values) from two Guids (2 128-bit values)
      g = Guid.NewGuid();
      _s0 = *(ulong*)&g;
      _s1 = *((ulong*)&g + 1);
  
      g = Guid.NewGuid();
      _s2 = *(ulong*)&g;
      _s3 = *((ulong*)&g + 1);
  }
  while ((_s0 | _s1 | _s2 | _s3) == 0); // at least one value must be non-zero
```

So I think using Interop.GetRandomBytes() instead of NewGuid() is all that's needed when invoking the default constructor - and all should be good.